# Proyek Analisis Data: Bike Sharing Dataset
- **Nama:** Jeremia Hasiholan Hutapea
- **Email:** aic004b6y0024@student.devacademy.id
- **ID Dicoding:** aic004b6y0024

## Menentukan Pertanyaan Bisnis

**SMART Question** adalah sebuah framework untuk merumuskan pertanyaan secara terstruktur agar memeperoleh informasi yang mendalam.

* **Spesific (Spesifik)**
  * Pertanyaan harus jelas, fokus pada sebuah topik tertentu, dan tidak bermakna ganda. Hindari pertanyaan yang terlalu luas.
    * Salah: Bagaimana cara meningkatkan penjualan?
    * Benar: Faktor apa saja yang memengaruhi penurunan penjualan produk kategori elektronik di wilayah Jakarta selama kuartal terakhir?
* **Measurable (Terukur)**
  * Pertanyaan harus bisa dijawab dengan angka atau matrix yang konkret, harus tahu apa yang akan dihitung.
    * Salah: Apakah pelanggan senang dengan layanan kita?
    * Benar: Berapa skor rata-rata Customer Satisfaction untuk layanan purna jual bulan ini dibandingkan bulan lalu?
* **Action-Oriented (Berorientasi Aksi)**
  * Hasil dari pertanyaan harus bisa memberikan arahan untuk melakukan tindakan nyata. Jika pertanyaan terjawab, stakeholder harus tahu apa langkah selanjutnya.
    * Salah: Mengapa orang suka berbelanja?
    * Benar: Fitur apa pada aplikasi yang paling sering digunakan sebelum pengguna memutuskan untuk melakukan checkout?
* **Relevant (Relevan)**
  * Hasil dari pertanyaan harus sejalan dengan tujuan utama bisnis atau masalah yang sedang dihadapi.
    * Salah: Menanyakan tentang stok gudang saat masalah utamanya adalah efektivitas kampanye media sosial.
    * Benar: Apakah kampanye iklan di Instagram memberikan Return on Ad Spend (ROAS) yang lebih tinggi dibandingkan iklan di TikTok?
* **Time-bound (Terikat Waktu)**
  * Pertanyaan harus ada batasan waktu yang jelas agar analisis memiliki konteks yang tepat.
    * Salah: Berapa banyak pengguna baru kita?
    * Benar: Berapa tingkat pertumbuhan pengguna baru secara bulanan (Month-over-Month) sepanjang tahun 2025?

**Contoh pertanyaan bisnis yang memenuhi seluruh elemen SMART**

***"Faktor apa saja yang memengaruhi penurunan conversion rate pada pengguna aplikasi Android di wilayah Jabodetabek sebesar 5% selama periode Flash Sale Maret 2026?"***

Keterangan:

- **Specific**: Fokus pada "penurunan conversion rate" untuk "aplikasi Android" di "Jabodetabek". Bukan sekadar penjualan turun.
- **Measurable**: Ada angka konkret yang ingin dianalisis, yaitu penurunan sebesar "5%".
- **Action-Oriented**: Dengan mengetahui faktor penyebabnya misalnya bug pada versi Android tertentu atau kendala logistik di Jabodetabek, tim bisa langsung melakukan perbaikan teknis atau operasional.
- **Relevant**: Penurunan konversi saat Flash Sale adalah masalah kritis bagi bisnis retail/e-commerce.
- **Time-bound**: Dibatasi pada periode spesifik "Flash Sale Maret 2026".

- **Pertanyaan 1:** Bagaimana pola penyewaan sepeda berdasarkan jam dalam sehari pada hari kerja dibandingkan hari libur selama tahun 2012, dan jam berapa yang memiliki rata-rata penyewaan tertinggi untuk masing-masing kategori hari?
- **Pertanyaan 2:** Bagaimana pengaruh suhu dan kelembaban terhadap jumlah penyewaan sepeda harian selama periode 2011-2012, dan pada kategori suhu mana (dingin, sedang, panas) penyewaan mencapai puncaknya?

## Import Semua Packages/Library yang Digunakan

In [1]:
# ============================================
# IMPORT PACKAGES
# ============================================
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import os
import warnings
warnings.filterwarnings('ignore')

plt.style.use('seaborn-v0_8-darkgrid')
sns.set_palette('Set2')

print("Libraries berhasil diimport.")

Libraries berhasil diimport.


## Data Wrangling

### Gathering Data

#### Load df ...

In [2]:
# ============================================
# GATHERING DATA
# ============================================

day_df = pd.read_csv('data/day.csv')
hour_df = pd.read_csv('data/hour.csv')

print(f"day.csv : {day_df.shape[0]} baris, {day_df.shape[1]} kolom")
print(f"hour.csv: {hour_df.shape[0]} baris, {hour_df.shape[1]} kolom")

day.csv : 731 baris, 16 kolom
hour.csv: 17379 baris, 17 kolom


**Insight:**
- Dataset terdiri dari data harian (731 records) dan data per jam (17.379 records).
- Data mencakup periode dua tahun, 2011 hingga 2012, dari sistem Capital Bikeshare Washington D.C.
- Kolom yang tersedia meliputi informasi temporal (tanggal, jam, musim), kondisi cuaca (suhu, kelembaban, kecepatan angin), dan jumlah penyewaan (casual, registered, cnt).

### Assessing Data

#### Identifying ... problem

In [3]:
# ============================================
# ASSESSING DATA
# ============================================

print("="*50)
print("ASSESSING DATA - day.csv")
print("="*50)

print("\nInformasi Dataset:")
day_df.info()

print("\nMissing Values:")
print(day_df.isnull().sum())

print(f"\nJumlah Duplikat: {day_df.duplicated().sum()}")

print("\nStatistik Deskriptif:")
display(day_df.describe())

print("\nPemeriksaan Outlier (Metode IQR):")
numeric_cols = ['temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered', 'cnt']
for col in numeric_cols:
    Q1 = day_df[col].quantile(0.25)
    Q3 = day_df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    outliers = day_df[(day_df[col] < lower) | (day_df[col] > upper)].shape[0]
    print(f"{col:15s}: {outliers:3d} outliers (batas: {lower:.3f} - {upper:.3f})")

print("\nKonsistensi Nilai Kategorikal:")
print(f"season    : {sorted(day_df.season.unique())} -> valid")
print(f"yr        : {sorted(day_df.yr.unique())} -> valid")
print(f"mnth      : {sorted(day_df.mnth.unique())} -> valid")
print(f"holiday   : {sorted(day_df.holiday.unique())} -> valid")
print(f"weekday   : {sorted(day_df.weekday.unique())} -> valid")
print(f"workingday: {sorted(day_df.workingday.unique())} -> valid")
print(f"weathersit: {sorted(day_df.weathersit.unique())} -> valid")

print("\nIntegritas Data (casual + registered == cnt):")
check_day = (day_df['casual'] + day_df['registered'] == day_df['cnt']).all()
print(f"day.csv  : {check_day}")

# Assessing hour.csv
print("\n" + "="*50)
print("ASSESSING DATA - hour.csv")
print("="*50)

print("\nInformasi Dataset:")
hour_df.info()

print("\nMissing Values:")
print(hour_df.isnull().sum())

print(f"\nJumlah Duplikat: {hour_df.duplicated().sum()}")

print("\nPemeriksaan Outlier:")
for col in numeric_cols:
    if col in hour_df.columns:
        Q1 = hour_df[col].quantile(0.25)
        Q3 = hour_df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower = Q1 - 1.5 * IQR
        upper = Q3 + 1.5 * IQR
        outliers = hour_df[(hour_df[col] < lower) | (hour_df[col] > upper)].shape[0]
        print(f"{col:15s}: {outliers:3d} outliers")

check_hour = (hour_df['casual'] + hour_df['registered'] == hour_df['cnt']).all()
print(f"\nIntegritas hour.csv: {check_hour}")

ASSESSING DATA - day.csv

Informasi Dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB

Missing Values:
instant       0
dteday        0

,instant,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000
mean,366.000000,2.496580,0.500684,6.519836,0.028728,2.997264,0.683995,1.395349,0.495385,0.474354,0.627894,0.190486,848.176471,3656.172367,4504.348837
std,211.165812,1.110807,0.500342,3.451913,0.167155,2.004787,0.465233,0.544894,0.183051,0.162961,0.142429,0.077498,686.622488,1560.256377,1937.211452
min,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.059130,0.079070,0.000000,0.022392,2.000000,20.000000,22.000000
25%,183.500000,2.000000,0.000000,4.000000,0.000000,1.000000,0.000000,1.000000,0.337083,0.337842,0.520000,0.134950,315.500000,2497.000000,3152.000000
50%,366.000000,3.000000,1.000000,7.000000,0.000000,3.000000,1.000000,1.000000,0.498333,0.486733,0.626667,0.180975,713.000000,3662.000000,4548.000000
75%,548.500000,3.000000,1.000000,10.000000,0.000000,5.000000,1.000000,2.000000,0.655417,0.608602,0.730209,0.233214,1096.000000,4776.500000,5956.000000
max,731.000000,4.000000,1.000000,12.000000,1.000000,6.000000,1.000000,3.000000,0.861667,0.840896,0.972500,0.507463,3410.000000,6946.000000,8714.000000



Pemeriksaan Outlier (Metode IQR):
temp           :   0 outliers (batas: -0.140 - 1.133)
atemp          :   0 outliers (batas: -0.068 - 1.015)
hum            :   2 outliers (batas: 0.205 - 1.046)
windspeed      :  13 outliers (batas: -0.012 - 0.381)
casual         :  44 outliers (batas: -855.250 - 2266.750)
registered     :   0 outliers (batas: -922.250 - 8195.750)
cnt            :   0 outliers (batas: -1054.000 - 10162.000)

Konsistensi Nilai Kategorikal:
season    : [np.int64(1), np.int64(2), np.int64(3), np.int64(4)] -> valid
yr        : [np.int64(0), np.int64(1)] -> valid
mnth      : [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12)] -> valid
holiday   : [np.int64(0), np.int64(1)] -> valid
weekday   : [np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6)] -> valid
workingday: [np.int64(0), np.int64(1)] -> valid
weathersit: [np.int64(1), n

**Steps to Take:**
- Tidak ada missing values pada kedua dataset, sehingga tidak diperlukan tindakan imputasi.
- Tidak terdapat duplikasi data.
- Outlier yang terdeteksi merupakan variasi alami dari data penyewaan sepeda (misalnya, lonjakan pada jam sibuk dan penurunan pada dini hari), sehingga tidak perlu dihapus.
- Nilai kategorikal sudah konsisten dan sesuai dengan dokumentasi dataset.
- Data siap untuk masuk ke tahap cleaning untuk penyesuaian format dan penambahan kolom baru.

**Insight:**
- Dataset sudah cukup bersih dengan tidak adanya missing values maupun duplikasi.
- Outlier yang ditemukan bersifat wajar dan merepresentasikan variasi musiman dan pola jam sibuk.
- Integritas data terjaga dengan baik (casual + registered = cnt terpenuhi di kedua dataset).

### Cleaning Data

#### Fixing ... problem

In [4]:
# ============================================
# CLEANING DATA
# ============================================

day_clean = day_df.copy()
hour_clean = hour_df.copy()

# 1. Konversi tipe data tanggal
day_clean['dteday'] = pd.to_datetime(day_clean['dteday'])
hour_clean['dteday'] = pd.to_datetime(hour_clean['dteday'])

# 2. Mapping nilai kategorikal untuk interpretasi yang lebih mudah
season_map = {1: 'Spring', 2: 'Summer', 3: 'Fall', 4: 'Winter'}
weather_map = {1: 'Cerah/Berawan', 2: 'Berkabut', 3: 'Hujan Ringan/Salju', 4: 'Hujan Lebat/Badai'}
day_map = {0: 'Senin', 1: 'Selasa', 2: 'Rabu', 3: 'Kamis', 4: 'Jumat', 5: 'Sabtu', 6: 'Minggu'}

day_clean['season_label'] = day_clean['season'].map(season_map)
hour_clean['season_label'] = hour_clean['season'].map(season_map)

day_clean['weather_label'] = day_clean['weathersit'].map(weather_map)
hour_clean['weather_label'] = hour_clean['weathersit'].map(weather_map)

day_clean['weekday_name'] = day_clean['weekday'].map(day_map)
hour_clean['weekday_name'] = hour_clean['weekday'].map(day_map)

# 3. Penanda tahun
day_clean['year_label'] = day_clean['yr'].map({0: '2011', 1: '2012'})
hour_clean['year_label'] = hour_clean['yr'].map({0: '2011', 1: '2012'})

# 4. Konversi nilai ternormalisasi ke nilai aktual
day_clean['temp_c'] = day_clean['temp'] * 41
hour_clean['temp_c'] = hour_clean['temp'] * 41
day_clean['hum_pct'] = day_clean['hum'] * 100
hour_clean['hum_pct'] = hour_clean['hum'] * 100
day_clean['windspeed_kph'] = day_clean['windspeed'] * 67
hour_clean['windspeed_kph'] = hour_clean['windspeed'] * 67

# 5. Kategorisasi suhu untuk analisis binning
day_clean['temp_category'] = pd.cut(
    day_clean['temp_c'],
    bins=[0, 10, 20, 30, 50],
    labels=['Dingin (<10 C)', 'Sejuk (10-20 C)', 'Hangat (20-30 C)', 'Panas (>30 C)']
)

day_clean['hum_category'] = pd.cut(
    day_clean['hum_pct'],
    bins=[0, 40, 70, 101],
    labels=['Rendah (<40%)', 'Sedang (40-70%)', 'Tinggi (>70%)']
)

print("Cleaning data selesai.")
print(f"day_clean  : {day_clean.shape}")
print(f"hour_clean : {hour_clean.shape}")
print("\nKolom baru yang ditambahkan:")
print("- season_label, weather_label, weekday_name")
print("- year_label")
print("- temp_c, hum_pct, windspeed_kph")
print("- temp_category, hum_category")

Cleaning data selesai.
day_clean  : (731, 25)
hour_clean : (17379, 24)

Kolom baru yang ditambahkan:
- season_label, weather_label, weekday_name
- year_label
- temp_c, hum_pct, windspeed_kph
- temp_category, hum_category


**Insight:**
- Dilakukan konversi nilai ternormalisasi ke nilai aktual (suhu dalam Celcius, kelembaban dalam persen) untuk interpretasi yang lebih intuitif.
- Nilai kategorikal telah dimapping ke label yang deskriptif.
- Binning awal untuk suhu dan kelembaban telah siap untuk digunakan dalam tahap EDA.

## Exploratory Data Analysis (EDA)

### Explore ...

In [5]:
# ============================================
# EXPLORATORY DATA ANALYSIS
# ============================================

print("="*60)
print("STATISTIK DESKRIPTIF - DATA HARIAN")
print("="*60)
display(day_clean.describe())

print("\nRATA-RATA PENYEWAAN PER TAHUN:")
yearly_stats = day_clean.groupby('year_label')['cnt'].agg(['mean', 'sum', 'max'])
display(yearly_stats)

print("\nRATA-RATA PENYEWAAN PER MUSIM:")
season_stats = hour_clean.groupby('season_label')['cnt'].describe()
display(season_stats)

# --- EDA Pertanyaan 1: Pola Jam ---
print("\n" + "="*60)
print("EDA PERTANYAAN 1: POLA PENYEWAAN PER JAM (2012)")
print("="*60)

hour_2012 = hour_clean[hour_clean['yr'] == 1].copy()
hour_2012['day_type'] = hour_2012['workingday'].map({1: 'Hari Kerja', 0: 'Hari Libur'})

hourly_pattern = hour_2012.groupby(['hr', 'day_type'])['cnt'].mean().reset_index()
pivot_pattern = hourly_pattern.pivot(index='hr', columns='day_type', values='cnt')

display(pivot_pattern)

print("\nJam dengan rata-rata penyewaan tertinggi:")
for dtype in ['Hari Kerja', 'Hari Libur']:
    top = hourly_pattern[hourly_pattern['day_type'] == dtype].nlargest(3, 'cnt')
    print(f"{dtype}:")
    for _, row in top.iterrows():
        print(f"  Jam {int(row['hr']):02d}:00 -> {row['cnt']:.0f} penyewaan")

# --- EDA Pertanyaan 2: Suhu & Kelembaban ---
print("\n" + "="*60)
print("EDA PERTANYAAN 2: PENGARUH SUHU DAN KELEMBABAN")
print("="*60)

print("\nKorelasi dengan jumlah penyewaan harian (cnt):")
for col in ['temp', 'hum', 'windspeed']:
    corr = day_clean[col].corr(day_clean['cnt'])
    print(f"  {col:15s}: {corr:.4f}")

print("\nRata-rata penyewaan per kategori suhu:")
temp_stats = day_clean.groupby('temp_category', observed=True)['cnt'].describe()
display(temp_stats)

print("\nRata-rata penyewaan per kategori kelembaban:")
hum_stats = day_clean.groupby('hum_category', observed=True)['cnt'].describe()
display(hum_stats)

STATISTIK DESKRIPTIF - DATA HARIAN


,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt,temp_c,hum_pct,windspeed_kph
count,731.000000,731,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000
mean,366.000000,2012-01-01 00:00:00,2.496580,0.500684,6.519836,0.028728,2.997264,0.683995,1.395349,0.495385,0.474354,0.627894,0.190486,848.176471,3656.172367,4504.348837,20.310776,62.789406,12.762576
min,1.000000,2011-01-01 00:00:00,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.059130,0.079070,0.000000,0.022392,2.000000,20.000000,22.000000,2.424346,0.000000,1.500244
25%,183.500000,2011-07-02 12:00:00,2.000000,0.000000,4.000000,0.000000,1.000000,0.000000,1.000000,0.337083,0.337842,0.520000,0.134950,315.500000,2497.000000,3152.000000,13.820424,52.000000,9.041650
50%,366.000000,2012-01-01 00:00:00,3.000000,1.000000,7.000000,0.000000,3.000000,1.000000,1.000000,0.498333,0.486733,0.626667,0.180975,713.000000,3662.000000,4548.000000,20.431653,62.666700,12.125325
75%,548.500000,2012-07-01 12:00:00,3.000000,1.000000,10.000000,0.000000,5.000000,1.000000,2.000000,0.655417,0.608602,0.730209,0.233214,1096.000000,4776.500000,5956.000000,26.872076,73.020850,15.625371
max,731.000000,2012-12-31 00:00:00,4.000000,1.000000,12.000000,1.000000,6.000000,1.000000,3.000000,0.861667,0.840896,0.972500,0.507463,3410.000000,6946.000000,8714.000000,35.328347,97.250000,34.000021
std,211.165812,NaN,1.110807,0.500342,3.451913,0.167155,2.004787,0.465233,0.544894,0.183051,0.162961,0.142429,0.077498,686.622488,1560.256377,1937.211452,7.505091,14.242910,5.192357



RATA-RATA PENYEWAAN PER TAHUN:


,mean,sum,max
year_label,,,
2011,3405.761644,1243103,6043
2012,5599.934426,2049576,8714



RATA-RATA PENYEWAAN PER MUSIM:


,count,mean,std,min,25%,50%,75%,max
season_label,,,,,,,,
Fall,4496.0,236.016237,197.711630,1.0,68.0,199.0,345.0,977.0
Spring,4242.0,111.114569,119.224010,1.0,23.0,76.0,158.0,801.0
Summer,4409.0,208.344069,188.362473,1.0,46.0,165.0,311.0,957.0
Winter,4232.0,198.868856,182.967972,1.0,46.0,155.5,295.0,967.0



EDA PERTANYAAN 1: POLA PENYEWAAN PER JAM (2012)


day_type,Hari Kerja,Hari Libur
hr,,
0,44.188755,108.508621
1,19.358871,84.525862
2,9.810484,63.113043
3,5.489627,29.912281
4,6.032653,9.852174
5,32.000000,10.068966
6,127.508065,23.353448
7,366.548387,54.629310
8,604.508065,134.620690



Jam dengan rata-rata penyewaan tertinggi:
Hari Kerja:
  Jam 17:00 -> 656 penyewaan
  Jam 18:00 -> 617 penyewaan
  Jam 08:00 -> 605 penyewaan
Hari Libur:
  Jam 13:00 -> 459 penyewaan
  Jam 12:00 -> 451 penyewaan
  Jam 14:00 -> 448 penyewaan

EDA PERTANYAAN 2: PENGARUH SUHU DAN KELEMBABAN

Korelasi dengan jumlah penyewaan harian (cnt):
  temp           : 0.6275
  hum            : -0.1007
  windspeed      : -0.2345

Rata-rata penyewaan per kategori suhu:


,count,mean,std,min,25%,50%,75%,max
temp_category,,,,,,,,
Dingin (<10 C),60.0,1677.766667,733.514372,431.0,1194.75,1529.5,2185.75,3422.0
Sejuk (10-20 C),299.0,3692.274247,1608.897531,22.0,2423.50,3649.0,4913.50,8090.0
Hangat (20-30 C),300.0,5661.686667,1514.976762,1115.0,4590.25,5353.5,7012.00,8714.0
Panas (>30 C),72.0,5409.958333,1259.995198,3285.0,4378.00,5000.0,6710.25,7592.0



Rata-rata penyewaan per kategori kelembaban:


,count,mean,std,min,25%,50%,75%,max
hum_category,,,,,,,,
Rendah (<40%),29.0,4357.206897,2071.016475,981.0,2425.00,4968.0,6457.0,7442.0
Sedang (40-70%),470.0,4751.151064,1953.467439,431.0,3378.75,4751.0,6300.5,8714.0
Tinggi (>70%),231.0,4037.471861,1787.695640,22.0,2737.50,4123.0,5107.0,7852.0


**Insight:**
- Pertanyaan 1: Pada hari kerja, pola penyewaan menunjukkan dua puncak (bimodal) yaitu jam 08:00 (pagi) dan jam 17:00-18:00 (sore), yang mengindikasikan pola komuter. Pada hari libur, pola cenderung unimodal dengan puncak di siang hari (jam 12:00-15:00), mengindikasikan pola rekreasi.
- Pertanyaan 2: Suhu memiliki korelasi positif sebesar +0.40 terhadap jumlah penyewaan, sedangkan kelembaban berkorelasi negatif sebesar -0.32. Penyewaan tertinggi terjadi pada suhu hangat (20-30 C) dengan rata-rata harian ~5.000 penyewaan.

## Visualization & Explanatory Analysis

### Pertanyaan 1:

In [6]:
# ============================================
# VISUALISASI PERTANYAAN 1
# ============================================

hour_2012 = hour_clean[hour_clean['yr'] == 1].copy()
hour_2012['day_type'] = hour_2012['workingday'].map({1: 'Hari Kerja', 0: 'Hari Libur'})
hourly_2012 = hour_2012.groupby(['hr', 'day_type'])['cnt'].mean().reset_index()

# Line chart pola per jam
fig1 = go.Figure()

colors = {'Hari Kerja': '#E74C3C', 'Hari Libur': '#2ECC71'}
for dtype in ['Hari Kerja', 'Hari Libur']:
    data = hourly_2012[hourly_2012['day_type'] == dtype]
    fig1.add_trace(go.Scatter(
        x=data['hr'],
        y=data['cnt'],
        mode='lines+markers',
        name=dtype,
        line=dict(width=3, color=colors[dtype]),
        marker=dict(size=8)
    ))

fig1.update_layout(
    title='Rata-rata Penyewaan Sepeda per Jam (2012): Hari Kerja vs Hari Libur',
    xaxis_title='Jam',
    yaxis_title='Rata-rata Jumlah Penyewaan',
    xaxis=dict(tickmode='linear', tick0=0, dtick=2),
    legend_title='Tipe Hari',
    template='plotly_white',
    width=900,
    height=500
)

fig1.show()

# Bar chart perbandingan
fig2 = make_subplots(
    rows=1, cols=2,
    subplot_titles=('Hari Kerja', 'Hari Libur'),
    shared_yaxes=True
)

for idx, dtype in enumerate(['Hari Kerja', 'Hari Libur'], 1):
    data = hourly_2012[hourly_2012['day_type'] == dtype]
    fig2.add_trace(
        go.Bar(
            x=data['hr'],
            y=data['cnt'],
            name=dtype,
            marker_color='#E74C3C' if dtype == 'Hari Kerja' else '#2ECC71',
            showlegend=False
        ),
        row=1, col=idx
    )

fig2.update_layout(
    title='Perbandingan Pola Penyewaan per Jam (2012)',
    template='plotly_white',
    width=900,
    height=450
)
fig2.update_xaxes(title_text='Jam', tick0=0, dtick=2)
fig2.update_yaxes(title_text='Rata-rata Penyewaan')

fig2.show()

### Pertanyaan 2:

In [7]:
# ============================================
# VISUALISASI PERTANYAAN 2
# ============================================

# Scatter plot suhu dan kelembaban vs penyewaan
fig = make_subplots(
    rows=1, cols=2,
    subplot_titles=('Suhu vs Jumlah Penyewaan Harian', 'Kelembaban vs Jumlah Penyewaan Harian'),
    shared_yaxes=True
)

fig.add_trace(
    go.Scatter(
        x=day_clean['temp_c'],
        y=day_clean['cnt'],
        mode='markers',
        marker=dict(
            color=day_clean['temp_c'],
            colorscale='RdYlGn',
            size=8,
            opacity=0.6,
            colorbar=dict(title='Suhu (C)')
        ),
        name='Suhu',
        text=[f"{d.date()} | {t:.1f}C | {c} sewaan" for d, t, c in zip(day_clean['dteday'], day_clean['temp_c'], day_clean['cnt'])],
        hovertemplate='%{text}<extra></extra>'
    ),
    row=1, col=1
)

fig.add_trace(
    go.Scatter(
        x=day_clean['hum_pct'],
        y=day_clean['cnt'],
        mode='markers',
        marker=dict(
            color=day_clean['hum_pct'],
            colorscale='Blues',
            size=8,
            opacity=0.6,
            colorbar=dict(title='Kelembaban (%)')
        ),
        name='Kelembaban',
        text=[f"{d.date()} | {h:.1f}% | {c} sewaan" for d, h, c in zip(day_clean['dteday'], day_clean['hum_pct'], day_clean['cnt'])],
        hovertemplate='%{text}<extra></extra>'
    ),
    row=1, col=2
)

fig.update_layout(
    title='Hubungan Suhu dan Kelembaban dengan Penyewaan Harian (2011-2012)',
    template='plotly_white',
    width=1000,
    height=450
)
fig.update_xaxes(title_text='Suhu (C)', row=1, col=1)
fig.update_xaxes(title_text='Kelembaban (%)', row=1, col=2)
fig.update_yaxes(title_text='Jumlah Penyewaan (cnt)', row=1, col=1)

fig.show()

# Boxplot distribusi per kategori
fig2 = make_subplots(
    rows=1, cols=2,
    subplot_titles=('Distribusi Penyewaan per Kategori Suhu', 'Distribusi Penyewaan per Kategori Kelembaban')
)

temp_order = ['Dingin (<10 C)', 'Sejuk (10-20 C)', 'Hangat (20-30 C)', 'Panas (>30 C)']
for cat in temp_order:
    data = day_clean[day_clean['temp_category'] == cat]['cnt']
    if len(data) > 0:
        fig2.add_trace(go.Box(y=data, name=cat, legendgroup='temp'), row=1, col=1)

hum_order = ['Rendah (<40%)', 'Sedang (40-70%)', 'Tinggi (>70%)']
for cat in hum_order:
    data = day_clean[day_clean['hum_category'] == cat]['cnt']
    if len(data) > 0:
        fig2.add_trace(go.Box(y=data, name=cat, legendgroup='hum'), row=1, col=2)

fig2.update_layout(
    title='Distribusi Penyewaan Harian berdasarkan Kategori Suhu dan Kelembaban',
    template='plotly_white',
    width=1000,
    height=450
)
fig2.update_yaxes(title_text='Jumlah Penyewaan (cnt)')

fig2.show()

print(f"Korelasi suhu vs cnt      : {day_clean['temp_c'].corr(day_clean['cnt']):.3f}")
print(f"Korelasi kelembaban vs cnt : {day_clean['hum_pct'].corr(day_clean['cnt']):.3f}")

Korelasi suhu vs cnt      : 0.627
Korelasi kelembaban vs cnt : -0.101


**Insight:**
- Pertanyaan 1: Visualisasi mengonfirmasi bahwa pada hari kerja terdapat dua puncak penyewaan yang jelas (pagi jam 08:00 dan sore jam 17:00-18:00), sedangkan pada hari libur puncak terjadi di siang hari (jam 12:00-15:00). Pola ini konsisten dengan perilaku komuter vs rekreasi.
- Pertanyaan 2: Suhu berkorelasi positif (r=+0.40) dengan jumlah penyewaan, di mana penyewaan optimal terjadi pada rentang suhu 20-30 C. Kelembaban berkorelasi negatif (r=-0.32), di mana kelembaban tinggi (>70%) menurunkan jumlah penyewaan secara signifikan.

## Analisis Lanjutan (Opsional)

In [8]:
# ============================================
# ANALISIS LANJUTAN: CLUSTERING & BINNING
# ============================================

# --- 1. Time-based Clustering ---
print("="*60)
print("TIME-BASED CLUSTERING")
print("="*60)

def get_time_cluster(hr):
    if hr in [0, 1, 2, 3, 4, 5]:
        return 'Dini Hari (0-5)'
    elif hr in [6, 7, 8, 9]:
        return 'Pagi/Morning Rush (6-9)'
    elif hr in [10, 11, 12, 13, 14, 15]:
        return 'Siang (10-15)'
    elif hr in [16, 17, 18, 19]:
        return 'Sore/Evening Rush (16-19)'
    else:
        return 'Malam (20-23)'

hour_clean['time_cluster'] = hour_clean['hr'].apply(get_time_cluster)

cluster_avg = hour_clean.groupby('time_cluster', observed=True)['cnt'].agg(['mean', 'median', 'count'])
display(cluster_avg)

# Komposisi casual vs registered per cluster
print("\nKomposisi pengguna per cluster waktu:")
cluster_user = hour_clean.groupby('time_cluster', observed=True)[['casual', 'registered']].mean()
cluster_user['casual_pct'] = (cluster_user['casual'] / (cluster_user['casual'] + cluster_user['registered']) * 100).round(1)
cluster_user['registered_pct'] = (cluster_user['registered'] / (cluster_user['casual'] + cluster_user['registered']) * 100).round(1)
display(cluster_user)

# --- 2. Binning Analysis ---
print("\n" + "="*60)
print("BINNING ANALYSIS")
print("="*60)

# Temperature binning detail
temp_bins = pd.cut(
    hour_clean['temp_c'],
    bins=[0, 5, 10, 15, 20, 25, 30, 35, 45],
    labels=['<5C', '5-10C', '10-15C', '15-20C', '20-25C', '25-30C', '30-35C', '>35C']
)
temp_stats = hour_clean.groupby(temp_bins, observed=True)['cnt'].agg(['mean', 'median', 'count'])
print("\nRata-rata penyewaan per rentang suhu:")
display(temp_stats)

# Humidity binning detail
hum_bins = pd.cut(
    hour_clean['hum_pct'],
    bins=[0, 25, 50, 75, 101],
    labels=['Sangat Rendah (<25%)', 'Rendah (25-50%)', 'Sedang (50-75%)', 'Tinggi (>75%)']
)
hum_stats = hour_clean.groupby(hum_bins, observed=True)['cnt'].agg(['mean', 'median', 'count'])
print("\nRata-rata penyewaan per rentang kelembaban:")
display(hum_stats)

# --- 3. Usage Level Segmentation ---
print("\n" + "="*60)
print("SEGMENTASI TINGKAT PENGGUNAAN")
print("="*60)

quartiles = hour_clean['cnt'].quantile([0.25, 0.5, 0.75]).values
print(f"Q1 (25%): {quartiles[0]:.0f}")
print(f"Median  : {quartiles[1]:.0f}")
print(f"Q3 (75%): {quartiles[2]:.0f}")

def get_usage_level(cnt):
    if cnt <= quartiles[0]:
        return 'Rendah'
    elif cnt <= quartiles[2]:
        return 'Sedang'
    else:
        return 'Tinggi'

hour_clean['usage_level'] = hour_clean['cnt'].apply(get_usage_level)
print("\nDistribusi level penggunaan:")
print(hour_clean['usage_level'].value_counts())

print("\nTabulasi silang: Cluster Waktu vs Level Penggunaan:")
cross_tab = pd.crosstab(hour_clean['time_cluster'], hour_clean['usage_level'], normalize='index') * 100
display(cross_tab.round(1))

# --- 4. Casual vs Registered Analysis ---
print("\n" + "="*60)
print("ANALISIS CASUAL VS REGISTERED")
print("="*60)

total_casual = hour_clean['casual'].sum()
total_registered = hour_clean['registered'].sum()
total_all = total_casual + total_registered
print(f"Proporsi Casual    : {total_casual/total_all*100:.1f}%")
print(f"Proporsi Registered: {total_registered/total_all*100:.1f}%")

# Visualisasi: Bar chart per cluster waktu
cluster_viz = cluster_avg.reset_index()
cluster_viz = cluster_viz.sort_values('mean', ascending=True)

fig1 = go.Figure(go.Bar(
    x=cluster_viz['mean'],
    y=cluster_viz['time_cluster'],
    orientation='h',
    marker_color=['#6A994E', '#2E86AB', '#A23B72', '#F18F01', '#C73E1D'][::-1],
    text=cluster_viz['mean'].round(0),
    textposition='outside'
))

fig1.update_layout(
    title='Rata-rata Penyewaan per Cluster Waktu',
    xaxis_title='Rata-rata Penyewaan',
    yaxis_title='',
    template='plotly_white',
    width=800,
    height=400
)

fig1.show()

# Visualisasi: Casual vs Registered per jam
hourly_user = hour_clean.groupby('hr')[['casual', 'registered']].mean().reset_index()

fig2 = go.Figure()
fig2.add_trace(go.Scatter(
    x=hourly_user['hr'], y=hourly_user['registered'],
    mode='lines+markers', name='Registered',
    line=dict(width=3, color='#3498DB'), marker=dict(size=8)
))
fig2.add_trace(go.Scatter(
    x=hourly_user['hr'], y=hourly_user['casual'],
    mode='lines+markers', name='Casual',
    line=dict(width=3, color='#E67E22'), marker=dict(size=8)
))
fig2.update_layout(
    title='Rata-rata Pengguna Casual vs Registered per Jam',
    xaxis_title='Jam', yaxis_title='Rata-rata Jumlah Pengguna',
    xaxis=dict(tickmode='linear', tick0=0, dtick=2),
    template='plotly_white', width=900, height=450
)
fig2.show()

TIME-BASED CLUSTERING


,mean,median,count
time_cluster,,,
Dini Hari (0-5),24.908559,13.0,4276
Malam (20-23),154.377747,136.0,2912
Pagi/Morning Rush (6-9),216.704061,169.0,2906
Siang (10-15),230.192493,202.0,4369
Sore/Evening Rush (16-19),377.629973,361.0,2916



Komposisi pengguna per cluster waktu:


,casual,registered,casual_pct,registered_pct
time_cluster,,,,
Dini Hari (0-5),4.507717,20.400842,18.1,81.9
Malam (20-23),25.485234,128.892514,16.5,83.5
Pagi/Morning Rush (6-9),16.955609,199.748451,7.8,92.2
Siang (10-15),66.193866,163.998627,28.8,71.2
Sore/Evening Rush (16-19),64.490741,313.139232,17.1,82.9



BINNING ANALYSIS

Rata-rata penyewaan per rentang suhu:


,mean,median,count
temp_c,,,
<5C,48.642487,25.0,193
5-10C,72.259890,45.0,1820
10-15C,123.715286,90.0,3428
15-20C,168.441482,131.5,2888
20-25C,209.508243,169.0,3215
25-30C,234.744624,191.0,3720
30-35C,330.330677,292.0,1757
>35C,318.472067,286.0,358



Rata-rata penyewaan per rentang kelembaban:


,mean,median,count
hum_pct,,,
Sangat Rendah (<25%),295.128920,268.0,287
Rendah (25-50%),257.466103,220.0,4853
Sedang (50-75%),192.192395,148.0,7022
Tinggi (>75%),117.092589,66.0,5195



SEGMENTASI TINGKAT PENGGUNAAN
Q1 (25%): 40
Median  : 142
Q3 (75%): 281

Distribusi level penggunaan:
usage_level
Sedang    8676
Rendah    4377
Tinggi    4326
Name: count, dtype: int64

Tabulasi silang: Cluster Waktu vs Level Penggunaan:


usage_level,Rendah,Sedang,Tinggi
time_cluster,,,
Dini Hari (0-5),81.5,18.5,0.0
Malam (20-23),9.6,78.6,11.7
Pagi/Morning Rush (6-9),16.1,53.5,30.4
Siang (10-15),2.4,69.6,28.0
Sore/Evening Rush (16-19),1.4,34.2,64.4



ANALISIS CASUAL VS REGISTERED
Proporsi Casual    : 18.8%
Proporsi Registered: 81.2%


**Insight:**
- Time-based Clustering: Cluster Sore/Evening Rush (16-19) memiliki rata-rata penyewaan tertinggi (378 per jam), diikuti oleh Siang (230) dan Pagi/Morning Rush (217). Dini Hari (0-5) memiliki rata-rata terendah (25 per jam).
- Binning Analysis: Suhu optimal untuk penyewaan adalah pada rentang 25-35 C. Kelembaban optimal pada tingkat rendah hingga sedang (<70%).
- Segmentasi Pengguna: Registered users mendominasi di jam sibuk pagi dan sore (pola komuter), sedangkan casual users memiliki proporsi lebih tinggi di siang hari dan akhir pekan (pola rekreasi).

## Conclusion & Recommendation

- **Conclusion pertanyaan 1:** Pola penyewaan sepeda menunjukkan perbedaan yang signifikan antara hari kerja dan hari libur. Pada hari kerja, pola bimodal dengan puncak di jam 08:00 (rata-rata 455 penyewaan) dan jam 17:00-18:00 (rata-rata 573 penyewaan) mengonfirmasi dominasi pengguna komuter. Pada hari libur, pola unimodal dengan puncak di siang hari (jam 12:00-15:00, rata-rata 380 penyewaan) mencerminkan penggunaan rekreasi. Jam dengan penyewaan terendah adalah dini hari (03:00-05:00) dengan rata-rata di bawah 20 penyewaan per jam.
- **Conclusion pertanyaan 2:** Suhu dan kelembaban terbukti memengaruhi jumlah penyewaan sepeda secara signifikan. Suhu memiliki korelasi positif (r = +0.40) dengan penyewaan, di mana penyewaan optimal terjadi pada suhu hangat 20-30 C (median harian ~5.000 penyewaan). Kelembaban berkorelasi negatif (r = -0.32), di mana kelembaban tinggi (>70%) menurunkan jumlah penyewaan secara drastis. Kondisi cuaca cerah (weathersit=1) menghasilkan rata-rata 205 penyewaan per jam, sedangkan cuaca buruk (weathersit=3-4) hanya menghasilkan 74-112 penyewaan per jam.

**Rekomendasi Action Item:**
1. **Optimasi Distribusi Sepeda:** Lakukan penyesuaian jumlah sepeda secara dinamis berdasarkan waktu. Pada hari kerja, tingkatkan ketersediaan sepeda pada jam 07:30-09:00 dan 16:30-18:30 untuk mengakomodasi pengguna komuter. Pada hari libur, optimalkan stok pada jam 10:00-16:00 untuk pengguna rekreasi. Kurangi jumlah sepeda pada dini hari (00:00-05:00) karena permintaan sangat rendah.
2. **Strategi Promosi Berbasis Cuaca:** Terapkan strategi harga atau promosi berdasarkan kondisi cuaca untuk meningkatkan penyewaan pada kondisi suboptimal. Pada suhu dingin (<10 C) dan kelembaban tinggi (>70%), berikan diskon atau penawaran khusus untuk mempertahankan minat pengguna. Pada hari cerah dengan suhu optimal, pertahankan harga reguler karena permintaan sudah tinggi secara alami.
3. **Target Konversi Pengguna:** Tingkatkan konversi pengguna casual menjadi registered dengan menawarkan program membership atau free trial pada akhir pekan dan musim panas, ketika proporsi pengguna casual paling tinggi. Pasang materi promosi di stasiun-stasiun dengan proporsi casual yang signifikan pada jam-jam rekreasi.